Docs: Add Gutenberg-specific JSDoc recommendations (#18920)
WordPress/gutenberg: docs/contributors/coding-guidelines.md

### Custom Types

Define custom types using the [JSDoc `@typedef` tag](https://jsdoc.app/tags-typedef.html).

A custom type should include a description, and should always include its base type.

Custom types should be named as succinctly as possible, while still retaining clarity of meaning and avoiding conflict with other global or scoped types. A `WP` prefix should be applied to all custom types. Avoid superfluous or redundant prefixes and suffixes (for example, a `Type` suffix, or `Custom` prefix). Custom types are not global by default, so a custom type does not need to be excessively specific to a particular package. However, they should be named with enough specificity to avoid ambiguity or name collisions when brought into the same scope as another type.

```js
/**
 * A block selection object.
 *
 * @typedef {Object} WPBlockSelection
 *
 * @property {string} clientId     A block client ID.
 * @property {string} attributeKey A block attribute key.
 * @property {number} offset       An attribute value offset, based on the rich
 *                                 text value.
 */
```

> Custom types can also be beneficial when describing a set of predefined options. While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids a situation where aligning tags while still respecting a maximum line length of 80 characters becomes challenging due to the space occupied by a literal values union.

Reviewer: ```suggestion
While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, it might quickly become challenging to align tags while still respecting a maximum line length of 80 characters. Using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids the issue created by the space occupied by a literal values union.
```

I'd be nice to split this sentence into smaller parts to make it easier to read. I don't know if my proposal makes it any better but it shows the idea.

Author: >I'd be nice to split this sentence into smaller parts to make it easier to read. I don't know if my proposal makes it any better but it shows the idea.

I went with something slightly different in 1bb94137d, but hopefully to the same effect.

```js
/**
 * Named breakpoint sizes.
 * 
 * @typedef {"huge"|"wide"|"large"|"medium"|"small"|"mobile"} WPBreakpoint
 */
```

### Importing and Exporting Types

Use the [TypeScript `import` function](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#import-types) to import type declarations from other files or third-party dependencies.

Since an imported type declaration can occupy an excess of the available line length and become verbose when referenced multiple times, you are encouraged to create an alias of the external type using a `@typedef` declaration at the top of the file, immediately following [the `import` groupings](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports).

```js

> /** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */

Reviewer: I think @dsifford raised this already, we should use either `'` or `"` in all places.

Author: > I think @dsifford raised this already, we should use either `'` or `"` in all places.

Updated in 88a12b1f3, and added additional guidelines in 4ba73de83.

```

Note that all custom types defined in another file can be imported.

> When considering which types should be made available from a WordPress package, the `@typedef` statements in the package's entrypoint script should be treated as effectively the same as its public API. It is important to be aware of this, both to avoid unintentionally exposing internal types on the public interface, and as the means by which to expose the public types of a project.

Reviewer: >  and as the means by which to expose the

This could be reworded to use simpler language, I'm thinking mostly for non-native speakers like myself.

Author: Thinking to go with "as a way to expose" vs. "as the means by which to expose". Better?

Reviewer: Yes, it works for me 👍

Author: > Thinking to go with "as a way to expose" vs. "as the means by which to expose". Better?

Updated in 9727cb489.

> When considering which types should be made available from a WordPress package, the `@typedef` statements in the package's entrypoint script should be treated as effectively the same as its public API. It is important to be aware of this, both to avoid unintentionally exposing internal types on the public interface, and as the means by which to expose the public types of a project.

Reviewer: I remember updating the spelling of entry point in the webpack dependencies plugin, can we align:

```suggestion
When considering which types should be made available from a WordPress package, the `@typedef` statements in the package's entry point script should be treated as effectively the same as its public API. It is important to be aware of this, both to avoid unintentionally exposing internal types on the public interface, and as the means by which to expose the public types of a project.
```

Author: >I remember updating the spelling of entry point in the webpack dependencies plugin, can we align:

Updated in 1f1a69a39.

```js
// packages/data/src/index.js

> /** @typedef {import('./registry').WPDataRegistry} WPDataRegistry */

Reviewer: Did you confirm it works seamlessly with VSC? I had some issues with it but I also installed VSC for the first time just before testing it 😛

Author: To make this work when referencing the types externally, it will require some extra changes, yes.

See also: #18927

As seen in the example there, it will start to work once that's in place.

```

In this snippet, the `@typedef` will support the usage of the previous example's `import('@wordpress/data')`.

#### External Dependencies

Many third-party dependencies will distribute their own TypeScript typings. For these, the `import` semantics should "just work".

![Working Example: `import` type](https://user-images.githubusercontent.com/1779930/70167742-62198800-1695-11ea-9c21-82a91d4a60e2.png)

If you use a [TypeScript integration](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Editor-Support) for your editor, you can typically see that this works if the type resolves to anything other than the fallback `any` type.

For packages which do not distribute their own TypeScript types, you are welcomed to install and use the [DefinitelyTyped](http://definitelytyped.org/) community-maintained types definitions, if one exists.

> ### Record Types

Reviewer: The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.

At the same time, I like the idea of encouraging custom types to be defined in the file to make it easier to follow.

Author: > The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.

Yeah, I'll see what I can do to improve. I agree especially on the syntax for functions. I was also hesitant about even mentioning custom generic types. I think it could be okay as it is, in that it doesn't go into too much depth, but includes a relevant link for those who might be interested in using them.

Author: > > The topic covered in this section is very complex. It's also because I see this syntax for the first time so it might be a good idea to include some links for folks to learn the basics.
> 
> Yeah, I'll see what I can do to improve. I agree especially on the syntax for functions. I was also hesitant about even mentioning custom generic types. I think it could be okay as it is, in that it doesn't go into too much depth, but includes a relevant link for those who might be interested in using them.

Added some references in def916547, and changed the text around generic types to try to clarify that it's intended for advanced use-cases.

```js
/**
 * An apiFetch middleware handler. Passed the fetch options, the middleware is
 * expected to call the `next` middleware once it has completed its handling.
 *
 * @typedef {(options:WPAPIFetchOptions,next:WPAPIFetchMiddleware)=>void} WPAPIFetchMiddleware
 */
```

### Documenting Examples

Because the documentation generated using the `@wordpress/docgen` tool will include `@example` tags if they are defined, it is considered a best practice to include usage examples for functions and components. This is especially important for documented members of a package's public API.

When documenting an example, use the markdown <code>\`\`\`</code> code block to demarcate the beginning and end of the code sample. An example can span multiple lines.

```js
/**
 * Given the name of a registered store, returns an object of the store's
 * selectors. The selector functions are been pre-bound to pass the current
 * state automatically. As a consumer, you need only pass arguments of the
 * selector, if applicable.
 *
 * @param {string} name Store name.
 *
 * @example
 * ```js
 * select( 'my-shop' ).getPrice( 'hammer' );
 * ```
 *

>  * @return {Object} Object containing the store's selectors.

Reviewer: Nit: It goes again the previous statement to avoid using `{Object}` :)

> When documenting a generic type such as `Object`, `Function`, `Promise`, etc., always include details about the expected record types.

Author: > Nit: It goes again the previous statement to avoid using `{Object}` :)

Good catch. I found myself needing to fix a few of the other examples I had pulled from real code 😄 In fact, with this one, I also already changed to remove the [reference to `wp.data.select` global](https://github.com/WordPress/gutenberg/blob/da0b0016869721b4844f8bd1a11e67aaae8e165d/packages/data/src/index.js#L83), since I'm not sure if this is really the direction we want to be promoting in examples for referencing the functions externally.

Author: > > Nit: It goes again the previous statement to avoid using `{Object}` :)
> 
> Good catch. I found myself needing to fix a few of the other examples I had pulled from real code 😄 In fact, with this one, I also already changed to remove the [reference to `wp.data.select` global](https://github.com/WordPress/gutenberg/blob/da0b0016869721b4844f8bd1a11e67aaae8e165d/packages/data/src/index.js#L83), since I'm not sure if this is really the direction we want to be promoting in examples for referencing the functions externally.

Corrected this and another error in b3c370026.

 */
```

### Documenting `@wordpress/element` (React) Components

When possible, all components should be implemented as [function components](https://reactjs.org/docs/components-and-props.html#function-and-class-components), using [hooks](https://reactjs.org/docs/hooks-intro.html) for managing component lifecycle and state.

Documenting a function component should be treated the same as any other function. The primary caveat in documenting a component is being aware that the function typically accepts only a single argument (the "props"), which may include many property members. Use the [dot syntax for parameter properties](https://jsdoc.app/tags-param.html#parameters-with-properties) to document individual prop types.

```js
/**
 * Renders the block's configured title as a string, or empty if the title
 * cannot be determined.
 *
 * @example
 *
 * ```jsx

>  * <BlockTitle clientId="afd1cb17-2c08-4e7a-91be-007ba7ddc3a1" />

Reviewer: The example should align with the props listed in the `@param`:
```suggestion
 * <BlockTitle name="my-plugin/my-block" />
```

or the other way around.

Author: > The example should align with the props listed in the `@param`:

Updated in 03985e79e.
